model/flow/factory/cluster_list.go

```diff
@@ -43,7 +43,7 @@ func NewClusterList(assignments flow.AssignmentList, collectors flow.IdentityLis
 			delete(lookup, participantID)
 
 			if i > 0 {
-				if !order.IdentifierCanonical(prev, participantID) {
+				if order.IdentifierCanonical(prev, participantID) > 0 {
```

Comment:
this line has a breaking change: 
Code to the left errors if two identities are equal. Code to the right does not error for equal entities. 
This is made to match the behaviour of other similar function. It also assumes the identities have no duplicates.

Reply:
The convention for canonical ordering is
* strictly monotonic increasing node IDs (lexicographical order).

   The strictness is important here, meaning that we do not allow duplicates. 

This code (including other places where we check for canonical ordering) is there as a check to guarantee that central conventions in the protocol are maintained. We check it here once and reject inputs that are not canonically ordered (including rejecting identity lists that contain duplicates). Each node locally verifying canonical ordering at the time the node locally constructs the cluster, we can guarantee that any code dealing with identities later does not have to worry about it. 

With this change, I think we are weakening correctness guarantees:
* previously, the node would reject identity lists with duplicates
* now we rely on other parts of the system to not break this convention. If that convention happens to be broken (e.g. bug or parts of the system being compromised), we won't detect it here resulting in undefined behaviour. 

Therefore, I would like to insist that _any check for canonical ordering must reject inputs with duplicated nodes_!

Reply:
reverted to original behaviour in https://github.com/onflow/flow-go/pull/5155/commits/a478ef83affbc3e005052300d10e462161a8426b (although redundant test is [here](https://github.com/onflow/flow-go/blob/tarak/update-go-exp/model/flow/factory/cluster_list.go#L25))